fix(HMS-2178): followup context fix for statuser

lzap opened the pull request:
So the previous fix we merged yesterday was not complete, I tested if it helped and unfortunately it did not so I kept digging. I have found a reason tho - when we receive kafka message, we extract identity, put it into a struct which is passed to worker and sending goroutines. We process everything and send it. There are problems.

We still see `Not found error from sources` and in order to find what is wrong, I need to see the whole logging transaction that can be correlated by msg_id or org_id. This is currently not possible:

![image](https://github.com/RHEnVision/provisioning-backend/assets/49752/4ca4cc96-06f1-48ce-8df9-7c1d31a2a1a3)

First, the goroutines which are processing have context previously named `ctx` and this context is used for logging. But this is the "global" context, the logger in that context does not have the msg_id, org_id and other required fields which are needed to correlate log records. It is supposed to only be used for context cancellation - e.g. when the service is shutting down and it should immediately break the loop and exit. Therefore I renamed this context to `cancelCtx` and it is not used, with two exceptions - when the timer expires and messages need to be sent (no message context available) and when the function needs to cancel (also there is no message so this is the only option).

The correct context we need to use is the message context, it has everything because the kafka consumer in the application prepares it and makes sure it is always a copy (see the previous PR https://github.com/RHEnVision/provisioning-backend/pull/606). We were discussing some time ago if it is good practice to pass context through channels and the opinion on Gopher slack was - it is possible, tho, we do not recommend to encapsulate only data that needs to be sent. This was true for identity, but now that we also want to pass logger around we have two data values: logger and identity. I think in this case it makes sense to pass copies of context around.

The 404 we see from sources is coming from the `clients.GetSourcesClient(ctx)` so something is wrong with that. We will see more info once logging is fixed, I believe that identity is incorrectly passed or something. This patch also adds trace logging of all sent kafka messages (headers, body) so we can investigate further the contents of messages on stage.

avitova commented on internal/kafka/avail_status_msg.go:
```diff
@@ -40,10 +41,10 @@ func genericMessage(ctx context.Context, m any, key string, topic string) (Gener
-			"x-rh-sources-org-id", id.Identity.OrgID,
-			"x-rh-sources-account-number", id.Identity.AccountNumber,
+			"Content-Type", "application/json",
+			"X-RH-Identity", identity.IdentityHeader(ctx),
```
By the way, the X-RH-Identity is encoded in some way? I can see that we are decoding it with `base64.StdEncoding.DecodeString` somewhere else.

lzap replied:
Check documentation of the IdentityHeader function, yes, this is a base64-encoded JSON.

avitova reviewed (commented):
I have some nitpicks, but otherwise LGTM.

lzap commented on internal/kafka/avail_status_msg.go:
```diff
@@ -40,10 +41,10 @@ func genericMessage(ctx context.Context, m any, key string, topic string) (Gener
 		// Keep headers written in lowercase to match sources comparison.
 		Headers: GenericHeaders(
-			"content-type", "application/json",
-			"x-rh-identity", identity.IdentityHeader(ctx),
```
Consistency of what exactly? Header names? We are not consistent indeed, HTTP headers [are not case sensitive](https://stackoverflow.com/questions/5258977/are-http-headers-case-sensitive) and it is common to use CamelCase for them. However, TIL that for HTTP/2 the lowercase is actually enforced. This will be done by the HTTP/2 client, but I will use lowercase from now on. Let me fix thi.

lzap commented:
Rebased, changed the HTTP headers to all lowercase to keep consistency if that was the concern. But should not matter at all technically. Also dropped some logging from the permission checks, I think it is a bit too verbose.

What is interesting is IQE tests failed for statuser, however, I can now clearly see in the logs that it works fine:

```
9:47AM TRC Received message with key: 5, topic: platform.provisioning.internal.availability-check, offset: 0, partition: 0 hostname=provisioning-backend-statuser-db99f846f-2v66s payload="{\"source_id\":\"5\"}" version=6588
9:47AM TRC Sources availability check for 5 account_number=0369233 hostname=provisioning-backend-statuser-db99f846f-2v66s msg_id=7567856eacca4bac32ab5ed15e93457b org_id=3340851 source_id=5 trace_id=b06e9d5ecfecf0d9a6ce6444363cadc3 version=6588
9:47AM TRC otel: HTTP GET duration=5.631881 hostname=provisioning-backend-statuser-db99f846f-2v66s http.flavor=1.1 http.method=GET http.url=http://sources-api-svc.ephemeral-mbkc0x.svc:8000/api/sources/v3.1/sources/5/authentications net.peer.name=sources-api-svc.ephemeral-mbkc0x.svc otel=true span_id=55a25ebed28dd2c7 span_id_parent=865f065bc9588392 trace_id=2fc4a3f7a03b068d6c4379396e543563 version=6588
9:47AM TRC {"data":[{"id":"5","authtype":"provisioning-arn","username":"arn:aws:iam::093942615996:role/redhat-provisioning-arn-role","availability_status":"in_progress","resource_type":"Application","resource_id":"5"}],"meta":{"count":1,"limit":100,"offset":0},"links":{"first":"/api/sources/v3.1/sources/5/authentications?limit=100\u0026offset=0","last":"/api/sources/v3.1/sources/5/authentications?limit=100\u0026offset=100"}}
 account_number=0369233 content_length=418 hostname=provisioning-backend-statuser-db99f846f-2v66s msg_id=7567856eacca4bac32ab5ed15e93457b org_id=3340851 response_trace=true source_id=5 status="200 OK" status_code=200 trace_id=b06e9d5ecfecf0d9a6ce6444363cadc3 version=6588
9:47AM TRC otel: GetAuthentication duration=5.957333 hostname=provisioning-backend-statuser-db99f846f-2v66s otel=true span_id=865f065bc9588392 trace_id=2fc4a3f7a03b068d6c4379396e543563 version=6588
9:47AM TRC Checking AWS source availability status 5 account_number=0369233 hostname=provisioning-backend-statuser-db99f846f-2v66s msg_id=7567856eacca4bac32ab5ed15e93457b org_id=3340851 source_id=5 trace_id=b06e9d5ecfecf0d9a6ce6444363cadc3 version=6588
9:47AM DBG Listing policies attached to the role client=ec2 hostname=provisioning-backend-statuser-db99f846f-2v66s version=6588
9:47AM DBG Parsing ARN to get a friendly role name client=ec2 hostname=provisioning-backend-statuser-db99f846f-2v66s version=6588
9:47AM DBG Listing attached role policies client=ec2 hostname=provisioning-backend-statuser-db99f846f-2v66s version=6588
9:47AM DBG Fetching policies for attached policies client=ec2 hostname=provisioning-backend-statuser-db99f846f-2v66s version=6588
9:47AM TRC Sending 1 source availability status messages (tick) hostname=provisioning-backend-statuser-db99f846f-2v66s messages=1 version=6588
9:47AM TRC Sending 1 messages to Kafka hostname=provisioning-backend-statuser-db99f846f-2v66s topic=platform.sources.status version=6588
9:47AM TRC Kafka message body="{\"resource_id\":\"5\",\"resource_type\":\"Application\",\"status\":\"available\",\"error\":null}" headers={"Content-Type":"application/json","X-RH-Identity":"eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6IjAzNjkyMzMiLCJvcmdfaWQiOiIzMzQwODUxIiwiaW50ZXJuYWwiOnsib3JnX2lkIjoiMzM0MDg1MSIsImF1dGhfdGltZSI6NjMwMH0sInVzZXIiOnsidXNlcm5hbWUiOiJqZG9lIiwiZW1haWwiOiJqZG9lQGFjbWUuY29tIiwiZmlyc3RfbmFtZSI6IkpvaG4iLCJsYXN0X25hbWUiOiJEb2UiLCJpc19hY3RpdmUiOnRydWUsImlzX29yZ19hZG1pbiI6dHJ1ZSwiaXNfaW50ZXJuYWwiOmZhbHNlLCJsb2NhbGUiOiJlbl9VUyIsInVzZXJfaWQiOiIifSwic3lzdGVtIjp7fSwiYXNzb2NpYXRlIjp7IlJvbGUiOm51bGwsImVtYWlsIjoiIiwiZ2l2ZW5OYW1lIjoiIiwicmhhdFVVSUQiOiIiLCJzdXJuYW1lIjoiIn0sIng1MDkiOnsic3ViamVjdF9kbiI6IiIsImlzc3Vlcl9kbiI6IiJ9LCJ0eXBlIjoiVXNlciIsImF1dGhfdHlwZSI6ImJhc2ljLWF1dGgifX0=","X-RH-Sources-Account-Number":"0369233","X-RH-Sources-Org-Id":"3340851","event_type":"availability_status"} hostname=provisioning-backend-statuser-db99f846f-2v66s topic=platform.sources.status version=6588
9:47AM DBG writing 1 messages to platform.sources.status (partition: 0) hostname=provisioning-backend-statuser-db99f846f-2v66s kafka=true version=6588
9:48AM DBG looking up offset of kafka reader for partition 0 of platform.provisioning.internal.availability-check: 1 hostname=provisioning-backend-statuser-db99f846f-2v66s kafka=true version=6588
9:49AM DBG looking up offset of kafka reader for partition 0 of platform.provisioning.internal.availability-check: 1 hostname=provisioning-backend-statuser-db99f846f-2v66s kafka=true version=6588
9:49AM TRC Received message with key: 6, topic: platform.provisioning.internal.availability-check, offset: 1, partition: 0 hostname=provisioning-backend-statuser-db99f846f-2v66s payload="{\"source_id\":\"6\"}" version=6588
9:49AM TRC Sources availability check for 6 account_number=0369233 hostname=provisioning-backend-statuser-db99f846f-2v66s msg_id=a88ec028a262dea788319fc45e2c88a5 org_id=3340851 source_id=6 trace_id=eb3648a059024473854318ad6f4f206c version=6588
9:49AM TRC otel: HTTP GET duration=2.446823 hostname=provisioning-backend-statuser-db99f846f-2v66s http.flavor=1.1 http.method=GET http.url=http://sources-api-svc.ephemeral-mbkc0x.svc:8000/api/sources/v3.1/sources/6/authentications net.peer.name=sources-api-svc.ephemeral-mbkc0x.svc otel=true span_id=d68f981bb0c9846e span_id_parent=1b556b913ff181be trace_id=ef651e8baee1098d58186795642152a3 version=6588
9:49AM TRC {"data":[{"id":"6","authtype":"provisioning_lighthouse_subscription_id","username":"16f7f67b-7f3c-455a-95c5-5a4f22e416de","availability_status":"in_progress","resource_type":"Application","resource_id":"6"}],"meta":{"count":1,"limit":100,"offset":0},"links":{"first":"/api/sources/v3.1/sources/6/authentications?limit=100\u0026offset=0","last":"/api/sources/v3.1/sources/6/authentications?limit=100\u0026offset=100"}}
 account_number=0369233 content_length=418 hostname=provisioning-backend-statuser-db99f846f-2v66s msg_id=a88ec028a262dea788319fc45e2c88a5 org_id=3340851 response_trace=true source_id=6 status="200 OK" status_code=200 trace_id=eb3648a059024473854318ad6f4f206c version=6588
9:49AM TRC otel: GetAuthentication duration=2.738541 hostname=provisioning-backend-statuser-db99f846f-2v66s otel=true span_id=1b556b913ff181be trace_id=ef651e8baee1098d58186795642152a3 version=6588
9:49AM TRC Checking Azure source availability status 6 account_number=0369233 hostname=provisioning-backend-statuser-db99f846f-2v66s msg_id=a88ec028a262dea788319fc45e2c88a5 org_id=3340851 source_id=6 trace_id=eb3648a059024473854318ad6f4f206c version=6588
9:49AM TRC Sending 1 source availability status messages (tick) hostname=provisioning-backend-statuser-db99f846f-2v66s messages=1 version=6588
9:49AM TRC Sending 1 messages to Kafka hostname=provisioning-backend-statuser-db99f846f-2v66s topic=platform.sources.status version=6588
9:49AM TRC Kafka message body="{\"resource_id\":\"6\",\"resource_type\":\"Application\",\"status\":\"available\",\"error\":null}" headers={"Content-Type":"application/json","X-RH-Identity":"eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6IjAzNjkyMzMiLCJvcmdfaWQiOiIzMzQwODUxIiwiaW50ZXJuYWwiOnsib3JnX2lkIjoiMzM0MDg1MSIsImF1dGhfdGltZSI6NjMwMH0sInVzZXIiOnsidXNlcm5hbWUiOiJqZG9lIiwiZW1haWwiOiJqZG9lQGFjbWUuY29tIiwiZmlyc3RfbmFtZSI6IkpvaG4iLCJsYXN0X25hbWUiOiJEb2UiLCJpc19hY3RpdmUiOnRydWUsImlzX29yZ19hZG1pbiI6dHJ1ZSwiaXNfaW50ZXJuYWwiOmZhbHNlLCJsb2NhbGUiOiJlbl9VUyIsInVzZXJfaWQiOiIifSwic3lzdGVtIjp7fSwiYXNzb2NpYXRlIjp7IlJvbGUiOm51bGwsImVtYWlsIjoiIiwiZ2l2ZW5OYW1lIjoiIiwicmhhdFVVSUQiOiIiLCJzdXJuYW1lIjoiIn0sIng1MDkiOnsic3ViamVjdF9kbiI6IiIsImlzc3Vlcl9kbiI6IiJ9LCJ0eXBlIjoiVXNlciIsImF1dGhfdHlwZSI6ImJhc2ljLWF1dGgifX0=","X-RH-Sources-Account-Number":"0369233","X-RH-Sources-Org-Id":"3340851","event_type":"availability_status"} hostname=provisioning-backend-statuser-db99f846f-2v66s topic=platform.sources.status version=6588
9:49AM DBG writing 1 messages to platform.sources.status (partition: 0) hostname=provisioning-backend-statuser-db99f846f-2v66s kafka=true version=6588
```

avitova reviewed (approved):
TY!

lzap commented on cmd/pbackend/statuser.go:
```diff
@@ -41,11 +40,9 @@ import (
 	SourceApplicationID string
-
-	Identity identity.Principal
 }
```
So instead of identity, we pass the whole context (with logger and identity).

lzap commented on cmd/pbackend/statuser.go:
```diff
@@ -264,22 +286,24 @@ func sendResults(ctx context.Context, batchSize int, tickDuration time.Duration)
+				err := kafka.Send(cancelCtx, messages...)
 				if err != nil {
 					logger.Warn().Err(err).Msg("Could not send source availability status messages (cancel)")
 				}
```
This was a mess of misusing context, now it has `cancelCtx` which is only used for context cancellation and where message is not available (e.g. shutting down the service). For everything else there is `messageCtx` which is copied to `ctx` variable so it is more natural to write and copy-paste code.

lzap commented on cmd/pbackend/statuser.go:
```diff
@@ -374,7 +398,9 @@ func statuser() {
-	go sendResults(cancelCtx, 1024, 5*time.Second)
+	// processing can be slowed down by configuration: send messages in maximum
+	// batch size of 8 messages or 4 seconds (what comes first)
+	go sendResults(cancelCtx, 8, 4*time.Second)
```
Because we implemented slight slow-down (1 second) the buffer of 1k messages makes zero sense now, this is more tight configuration.

lzap commented on internal/kafka/avail_status_msg.go:
```diff
@@ -32,6 +32,7 @@ func genericMessage(ctx context.Context, m any, key string, topic string) (Gener
 	}
 
+	// This will panic when identity was not present in the context (no error handling possible)
 	id := identity.Identity(ctx)
```
This should improve it: https://github.com/RedHatInsights/platform-go-middlewares/pull/45

lzap commented on internal/kafka/kafka.go:
```diff
@@ -263,6 +263,16 @@ func (b *kafkaBroker) Send(ctx context.Context, messages ...*GenericMessage) err
+				Dict("headers", dict).
+				Msg("Kafka message")
+		}
 	}
```
All kafka messages are now logged including headers.

lzap commented on internal/kafka/source_result.go:
```diff
@@ -14,18 +12,12 @@ const (
+	Status             StatusType      `json:"status"`
+	Err                error           `json:"-"` // Sources do not support error field
+	MissingPermissions []string        `json:"-"` // Sources do not support reason field
 }
```
Before the patch, we were serializing "err" field, this changed it. Sources does not support this field, so let’s not send it.

lzap commented on internal/logging/zerolog.go:
```diff
@@ -58,7 +58,7 @@ func stdoutWriter(truncate bool) io.Writer {
 		NoColor:    config.InEphemeralClowder(),
-		TimeFormat: time.Kitchen,
+		TimeFormat: "15:04:05",
 	}
```
This helps with debugging, seconds were not present and I needed to see how long it took.
